### PR TITLE
Implement Either type as object

### PR DESCRIPTION
**Pull request by DzyubSpirit**
Refs: #31

### CODE REVIEW

File: lib/functional.js [hunk 1]
```diff
@@ -19,4 +19,89 @@ module.exports = (api) => {
     value !== undefined && value !== null ? fn(value) : defVal
   );
 
+  // Either can be used for saving either left type or right type value.
```
Comment (tshemsedinov) [inline lib/functional.js]: Description duplicated 2 times

File: lib/functional.js [hunk 2]
```diff
@@ -19,4 +19,89 @@ module.exports = (api) => {
...
+  // Left type is commonly used for additional information like error.
+  //
+  function Either(
+    // Either can be used for saving either left type or right type value.
+    // Left type is commonly used for additional information like error.
+    left, // left type value
+    right // right type value
+    // Returns tuple containing left type value or null as first element
+    // and right type value or undefined as second element
+  ) {
```
Comment (tshemsedinov) [inline lib/functional.js]:
```
this.isLeft = left;
this.val = left ? left : right;
```

Comment (belochub) [inline lib/functional.js]: @tshemsedinov, why not make second line like this?
```js
this.val = left || right;
```

File: lib/functional.js [hunk 3]
```diff
@@ -19,4 +19,89 @@ module.exports = (api) => {
...
+    e2 // second either value
+    // Returns result of appending two either values
+  ) => (e1[0] ? e1 : e2);
+
+  api.common.altConcat = (
+    // Alternative concatenation for either values
+    ...args // arguments for concatenation
+    // Returns result of concatenation of args
+  ) => {
+    switch (args.length) {
```
Comment (tshemsedinov) [inline lib/functional.js]: I'd prefer not using `switch` here, better `if (args.length === 0) return new Error('...');`

File: lib/functional.js [hunk 4]
```diff
@@ -19,4 +19,89 @@ module.exports = (api) => {
...
+      }
+    }
+  };
+
+  api.common.altConcatLazy = (
+    // Alternative lazy concatenation for either values
+    ...args // array of functions returning either values
+    // Returns result of concatenation of either values
+  ) => {
+    switch (args.length) {
```
Comment (tshemsedinov) [inline lib/functional.js]: Use `if` instead of `switch`

File: lib/functional.js (hunk 2 above)
Comment (tshemsedinov) [inline lib/functional.js]: @belochub sure! :+1:

**Comment by DzyubSpirit** [comment]
@tshemsedinov It's seems impossible to implement `altAppend` using closures because then `altAppend` function has to be declared in both `left1`, `right1` scope and `left2`, `right2` scope. It rule can be broken after implementing `getLeft`/`getRight`.
But I think it would be useless because encapsulation is good for systems with complex internal logic but isn't same good for simple structures like Either or Point(x, y).

File: lib/functional.js (hunk 2 above)
Comment (DzyubSpirit) [inline lib/functional.js]: @tshemsedinov, I am sorry it's my bad. I made if checking for a reason but forgot write good condition. I want that Either can save values of primitive types (not only errors) as left value. That wish needs complex checking of error.
```javascript
const [left, right] = api.common.makeEither(0); // left === 0, right === undefined
// Common checking
if (left) {
  console.log('left value');
} else {
  console.log('right value');
}
// prints 'right value'
// Complex checking
if (left !== null && left !== undefined) {
  console.log('left left');
} else {
  console.log('right value');
}
// prints 'left value'
```
That also shows hash implementation advantage. With tuple error checking has to be done every time. But with hash error checking is done once in constructor and then checking simplifies to checking `isLeft` flag

Comment (tshemsedinov) [inline lib/functional.js]:
```
this.isLeft = left !== null && left !== undefined;
this.val = this.isLeft ? left : right;
```

**Pull request closed without merging**